### ISSUE TITLE
Achieving test correlations reported in the paper

### ISSUE THREAD

**Issue by andrewj-jung** [johahi/borzoi-pytorch#17]
Hi! First of all, thanks for all the awesome work. It is great to have a PyTorch version of Borzoi, rather than dealing with TF! 

I have been trying to reproduce the test correlations reported in the paper, but correlations I get are much lower. I am wondering how was the test correlations computed? I've been using `fold3`, as done in the paper, and computing Pearson between the final model predictions after soft plus and target values from TFRecords. Did you have to filter out some examples, or the entire test split was used? 

Thank you so much for your help in advance and looking forward to your response!

**Comment by johahi** [comment]
Hi Andrew, thanks for your interest!
How much lower are your correlations?
For profile correlations, I first concatenated all predictions (fwd + rev_comp averaged) and targets along the genome for the entire test fold 3 to get `num_tracks x (num_samples_in_fold_3 * num_bins)`, without removing the squashed-scale and then calculated the Pearson corr per track.  I cast all values to float32. You can use torchmetrics for a "rolling" Pearson R if you don't have enough memory, it gives very similar results (see below).
This is an excerpt from my code, it's a bit ugly though:
```python
for i, [targets, input_ids]  in tqdm(enumerate(test_loader), disable = False):
    targets = targets.cuda(non_blocking=True).float()
    used_folds, input_ids = digest_input_ids(input_ids)
    inputs, shift_augs, rc_augs = get_inputs_and_augs(input_ids,ds,used_folds) # returns (bs, 524288,4), (0,0), False
    inputs = inputs.permute(0,2,1).to(device, non_blocking=True) 
    with torch.autocast("cuda"),torch.no_grad(), torch.inference_mode():
            outputs = flashzoi(inputs, is_human = True).permute(0,2,1)
            outputs_rev_comp = flashzoi(torch.flip(inputs.permute(0,2,1), (-1, -2)).permute(0,2,1), is_human = True).permute(0,2,1).flip(dims=[1])
            outputs_rev_comp = fix_rev_comp2(outputs_rev_comp,fix_indices_tensor_human) # fixes strand of rev_comp'ed tracks
    aggregated_outputs = (outputs + outputs_rev_comp)/2
    aggretated_targets =  targets
    val_loss += poisson_multinomial_torch(aggregated_outputs.float().detach(),aggretated_targets).item()
    pearson.update(aggregated_outputs.reshape(-1,7611).float(), aggretated_targets.reshape(-1,7611).float())
all_corrs_torch = pearson.compute()
```
Let me know if that helps!

**Comment by andrewj-jung** [comment]
Hi Johannes, really appreciate quick response! Following your advice I was able to get correlations close to what you reported on your paper. Thanks again! 

Previously, I was averaging Pearson computed per examples, and the results I get were a lot lower (although I just realized there could have been numerical instability since many examples have low variance in track values...).

I have one quick follow up question. It seems like most values are close to zeros and wondering if you think this could be inflating Pearson? In the Borzoi paper, they seem to do something quite complicated by only taking regions overlapping genes, and wondering if you have tried something like that and see differences?

**Comment by johahi** [comment]
Nice to hear! Took me a bit to reverse engineer this out back then... I agree the bin-level Pearson could be somewhat inflated, and shouldn't be the only metric, but as you say there are other metrics in the paper. I have replicated [Fig 1c](https://github.com/calico/borzoi-paper/blob/main/analysis/test_expression/plot_expr_histo_bins.ipynb) for borzoi-pytorch. 

> wondering if you have tried something like that and see differences?

I assume you mean differences between Flashzoi and Borzoi? I checked Fig 1c (in the Flashzoi-preprint) and performed an analysis similar to Fig 1d/e (actually more like we do in scooby Fig 2d,e): both look similar, slightly better for Flashzoi iirc. And variant effect pred. with Flashzoi is on par as well.

**Comment by andrewj-jung** [comment]
Gotcha! Thank you so much Johannes for all the help!
